pkg/util/util.go

```diff
@@ -22,3 +30,64 @@ func StringMatchRegexSlice(s string, re []string) bool {
 	}
 	return false
 }
+
+// ReadPassword reads a password from the terminal and restores the terminal state on interruption
+func ReadPassword(fd int) ([]byte, error) {
+	// Store and restore the terminal status on interruptions to
+	// avoid that the terminal remains in the password state
+	// This is necessary as for https://github.com/golang/go/issues/31180
+
+	/* WARNING: This commit contains two suggestions and should as such not be commited !! */
+	reemitSolution := true
+
+	/* Suggestion 1: Re-emit the interrupt signal, but restore the terminal in-between*/
+	if reemitSolution {
+		oldState, _ := term.GetState(fd)
+		interruptChannel := make(chan os.Signal, 1)
+		signal.Notify(interruptChannel, os.Interrupt)
+		defer signal.Stop(interruptChannel)
+		defer close(interruptChannel)
+
+		go func() {
+			for range interruptChannel {
+				if oldState != nil {
+					term.Restore(fd, oldState)
+				}
+				signal.Stop(interruptChannel)
+				syscall.Kill(syscall.Getpid(), syscall.SIGINT)
+			}
+		}()
+		return term.ReadPassword(fd)
+	} else {
+		/* Suggestion 2: Gracefully handle the interrupt signal, but leave the terminal input blocked */
+
+		type ByteBuffer struct {
+			buf []byte
+			err error
+		}
+
+		oldState, _ := term.GetState(fd)
+		interruptChannel := make(chan os.Signal, 1)
+		signal.Notify(interruptChannel, os.Interrupt)
+		defer signal.Stop(interruptChannel)
+		defer close(interruptChannel)
+
+		// Read from the file descriptor or wait for the interrupt signal
+		inputChannel := make(chan ByteBuffer, 1)
+		defer close(inputChannel)
+		go func() {
+			var buf ByteBuffer
+			buf.buf, buf.err = term.ReadPassword(fd)
+			inputChannel <- buf
```

Comment:
Suggestion 2:

With this, `term.ReadPassword` continues running, consuming resources (meh), but potentially interfering: it continues to consume data from stdin, and it can eventually return and change the terminal state at any future time.

I can’t see how we can terminate the current implementation: `fd` must be a terminal (so that ioctls work); and I can’t see that a pending `read(3)` would be interrupted by a `close(3)`.

---

A downside of turning the signal into an error is that the exit status of the process will not be traditionally correct. (In principle, we could turn it into a recognizable error, and turn that error into a signal at the very top of Podman.)

Reply:
Yes. The only way of solving this would be to change `term.ReadPassword` to handle interrupts itself gracefully.